From dfbe9988b51f830c544db452c5e95c6c6547f762 Mon Sep 17 00:00:00 2001
From: Nathan Chancellor <nathan@kernel.org>
Date: Thu, 16 Mar 2023 15:40:45 -0700
Subject: [PATCH 05/13] REPORTED: Revert "efi: libstub: Use relocated version
 of kernel's struct screen_info"

This reverts commit ef3efc2af044f6da5bb8c55e99f2398081d99c09.

Link: https://github.com/ClangBuiltLinux/linux/issues/1819
Signed-off-by: Nathan Chancellor <nathan@kernel.org>
---
 drivers/firmware/efi/libstub/efi-stub-entry.c | 18 ------------------
 drivers/firmware/efi/libstub/efi-stub.c       |  9 +++++++++
 drivers/firmware/efi/libstub/screen_info.c    |  7 +++++++
 3 files changed, 16 insertions(+), 18 deletions(-)

diff --git a/drivers/firmware/efi/libstub/efi-stub-entry.c b/drivers/firmware/efi/libstub/efi-stub-entry.c
index 629840696ea0..5245c4f031c0 100644
--- a/drivers/firmware/efi/libstub/efi-stub-entry.c
+++ b/drivers/firmware/efi/libstub/efi-stub-entry.c
@@ -5,19 +5,6 @@
 
 #include "efistub.h"
 
-#ifndef CONFIG_ARM
-static struct screen_info *si;
-
-struct screen_info *alloc_screen_info(void)
-{
-	return si;
-}
-
-void free_screen_info(struct screen_info *si)
-{
-}
-#endif
-
 /*
  * EFI entry point for the generic EFI stub used by ARM, arm64, RISC-V and
  * LoongArch. This is the entrypoint that is described in the PE/COFF header
@@ -69,11 +56,6 @@ efi_status_t __efiapi efi_pe_entry(efi_handle_t handle,
 		return status;
 	}
 
-#ifndef CONFIG_ARM
-	/* point si to the relocated copy of struct screen_info */
-	si = (void *)&screen_info + image_addr - (unsigned long)image->image_base;
-#endif
-
 	status = efi_stub_common(handle, image, image_addr, cmdline_ptr);
 
 	efi_free(image_size, image_addr);
diff --git a/drivers/firmware/efi/libstub/efi-stub.c b/drivers/firmware/efi/libstub/efi-stub.c
index c4b9eccad0f1..2955c1ac6a36 100644
--- a/drivers/firmware/efi/libstub/efi-stub.c
+++ b/drivers/firmware/efi/libstub/efi-stub.c
@@ -47,6 +47,15 @@
 static u64 virtmap_base = EFI_RT_VIRTUAL_BASE;
 static bool flat_va_mapping = (EFI_RT_VIRTUAL_OFFSET != 0);
 
+struct screen_info * __weak alloc_screen_info(void)
+{
+	return &screen_info;
+}
+
+void __weak free_screen_info(struct screen_info *si)
+{
+}
+
 static struct screen_info *setup_graphics(void)
 {
 	efi_guid_t gop_proto = EFI_GRAPHICS_OUTPUT_PROTOCOL_GUID;
diff --git a/drivers/firmware/efi/libstub/screen_info.c b/drivers/firmware/efi/libstub/screen_info.c
index b2637420fe53..8e76a8b384ba 100644
--- a/drivers/firmware/efi/libstub/screen_info.c
+++ b/drivers/firmware/efi/libstub/screen_info.c
@@ -15,7 +15,14 @@
  * early, but it only works if the EFI stub is part of the core kernel image
  * itself. The zboot decompressor can only use the configuration table
  * approach.
+ *
+ * In order to support both methods from the same build of the EFI stub
+ * library, provide this dummy global definition of struct screen_info. If it
+ * is required to satisfy a link dependency, it means we need to override the
+ * __weak alloc and free methods with the ones below, and those will be pulled
+ * in as well.
  */
+struct screen_info screen_info;
 
 static efi_guid_t screen_info_guid = LINUX_EFI_SCREEN_INFO_TABLE_GUID;
 
-- 
2.40.0

